src/Type/Generic/TemplateTypeTrait.php

```diff
@@ -270,6 +273,17 @@ public function getVariance(): TemplateTypeVariance
 		return $this->variance;
 	}
 
+	private function shouldInferTemplatesOn(IntersectionType $type): bool
```

Comment:
This method currently `return false` in all cases

Reply:
Good catch, thanks! It turns out this whole clause _appears_ to be causing the issue, so I've removed it altogether.